lib/ts/recipe/passwordless/components/features/signInAndUp/index.tsx

```diff
@@ -0,0 +1,257 @@
+                } else {
+                    contactInfo = formatPhoneNumberIntl(input.phoneNumber);
+                }
+                const res = await this.props.recipe.recipeImpl.createCode(input);
+                if (res.status === "OK") {
+                    const loginAttemptInfo = {
+                        ...res,
+                        lastResend: new Date().getTime(),
+                        contactMethod: this.props.recipe.config.contactMethod,
+                        contactInfo,
+                    };
+                    await this.props.recipe.recipeImpl.setLoginAttemptInfo(loginAttemptInfo);
+                    this.setState((os) => ({
+                        ...os,
+                        error: undefined,
+                        loginAttemptInfo,
+                    }));
+                } else if (res.status === "GENERAL_ERROR") {
+                    this.setState((os) => ({
+                        ...os,
+                        error: res.message,
+                        loginAttemptInfo: undefined,
+                    }));
+                }
+                return res;
+            },
+            resendCode: async (input) => {
+                if (!this.state.loginAttemptInfo) {
+                    // This should never happen, but it makes TS happy
+                    throw new Error("Resend without loginAttemptInfo");
+                }
+                const res = await this.props.recipe.recipeImpl.resendCode(input);
+                if (res.status === "OK") {
+                    const loginAttemptInfo = {
+                        ...this.state.loginAttemptInfo,
+                        lastResend: new Date().getTime(),
+                    };
+                    await this.props.recipe.recipeImpl.setLoginAttemptInfo(loginAttemptInfo);
+                    this.setState((os) => ({
+                        ...os,
+                        error: undefined,
+                        loginAttemptInfo,
+                    }));
+                } else if (res.status === "RESTART_FLOW_ERROR") {
+                    await this.props.recipe.recipeImpl.clearLoginAttemptInfo();
+
+                    this.setState((os) => ({
+                        ...os,
+                        error: SIGN_IN_UP_RESEND_RESTART_FLOW_ERROR,
+                        loginAttemptInfo: undefined,
+                    }));
+                } else if (res.status === "GENERAL_ERROR") {
+                    this.setState((os) => ({
+                        ...os,
+                        error: res.message,
+                        loginAttemptInfo: undefined,
+                    }));
+                }
+                return res;
+            },
+
+            consumeCode: async (input) => {
+                const res = await this.props.recipe.recipeImpl.consumeCode(input);
+                if (res.status === "RESTART_FLOW_ERROR") {
+                    await this.props.recipe.recipeImpl.clearLoginAttemptInfo();
```

Comment:
Why not call `clearLoginAttemptInfo` inside `recipeImpl. consumeCode` itself?

Reply:
We do not want to clear on `RESTART_FLOW_ERROR`: in the OTP + Link flow we can still continue if consuming the linkCode fails. We can check which the flow and what kind of consume ran and do the cleanup based on that, but I thought this kind of logic was supposed to be in the feature component, so the user could override it.